src/hotspot/share/utilities/intn_t.hpp

```diff
@@ -0,0 +1,141 @@
+  constexpr bool operator<(intn_t o) const { return int(*this) < int(o); }
+  constexpr bool operator>(intn_t o) const { return int(*this) > int(o); }
+  constexpr bool operator<=(intn_t o) const { return int(*this) <= int(o); }
+  constexpr bool operator>=(intn_t o) const { return int(*this) >= int(o); }
+};
+
+template <unsigned int n>
+unsigned count_leading_zeros(uintn_t<n>);
+
+template <unsigned int n>
+class uintn_t {
+  static_assert(n > 0 && n <= 8, "should not be larger than char");
+
+private:
+  uint _v;
+
+  constexpr static uint _mask = (1 << n) - 1;
+
+  friend class intn_t<n>;
+
+  friend unsigned count_leading_zeros<n>(uintn_t<n>);
+
+public:
+  explicit constexpr uintn_t(int v) : _v(v) {}
+  constexpr uintn_t() : _v(0) {}
+  constexpr uintn_t(const uintn_t&) = default;
+  constexpr uintn_t& operator=(const uintn_t&) = default;
+  explicit constexpr uintn_t(intn_t<n> v) : _v(v._v) {}
+  explicit constexpr operator uint() const { return _v & _mask; }
+
+  constexpr static int min = 0;
+  constexpr static int max = _mask;
+  static_assert(min < max, "");
+
+  constexpr bool operator==(uintn_t o) const { return (_v & _mask) == (o._v & _mask); }
+  constexpr bool operator!=(uintn_t o) const { return (_v & _mask) != (o._v & _mask); }
+  constexpr bool operator<(uintn_t o) const { return (_v & _mask) < (o._v & _mask); }
+  constexpr bool operator>(uintn_t o) const { return (_v & _mask) > (o._v & _mask); }
+  constexpr bool operator<=(uintn_t o) const { return (_v & _mask) <= (o._v & _mask); }
+  constexpr bool operator>=(uintn_t o) const { return (_v & _mask) >= (o._v & _mask); }
+  constexpr uintn_t operator+(uintn_t o) const { return uintn_t(_v + o._v); }
+  constexpr uintn_t operator-(uintn_t o) const { return uintn_t(_v - o._v); }
+  constexpr uintn_t operator&(uintn_t o) const { return uintn_t(_v & o._v); }
+  constexpr uintn_t operator|(uintn_t o) const { return uintn_t(_v | o._v); }
+  constexpr uintn_t operator^(uintn_t o) const { return uintn_t(_v ^ o._v); }
+  constexpr uintn_t operator>>(unsigned int s) const { return uintn_t((_v & _mask) >> s); }
+  constexpr uintn_t operator<<(unsigned int s) const { return uintn_t(_v << s); }
+  constexpr uintn_t operator~() const { return uintn_t(~_v); }
+  constexpr uintn_t operator-() const { return uintn_t(-_v); }
+  constexpr uintn_t& operator|=(uintn_t o) { _v |= o._v; return *this; }
+};
+
+template <unsigned int n>
+constexpr intn_t<n>::intn_t(uintn_t<n> v) : _v(v._v) {}
+
+namespace std {
+
+template <unsigned int n>
+class numeric_limits<intn_t<n>> {
+public:
+  constexpr static intn_t<n> min() { return intn_t<n>(intn_t<n>::min); }
+  constexpr static intn_t<n> max() { return intn_t<n>(intn_t<n>::max); }
+};
+
+template <unsigned int n>
+class numeric_limits<uintn_t<n>> {
+public:
+  constexpr static uintn_t<n> min() { return uintn_t<n>(uintn_t<n>::min); }
+  constexpr static uintn_t<n> max() { return uintn_t<n>(uintn_t<n>::max); }
+};
+
+}
```

Comment:
Could use indentation to make clear where the namespace starts and ends. Or at least a comment like this:
```suggestion
namespace std {

template <unsigned int n>
class numeric_limits<intn_t<n>> {
public:
  constexpr static intn_t<n> min() { return intn_t<n>(intn_t<n>::min); }
  constexpr static intn_t<n> max() { return intn_t<n>(intn_t<n>::max); }
};

template <unsigned int n>
class numeric_limits<uintn_t<n>> {
public:
  constexpr static uintn_t<n> min() { return uintn_t<n>(uintn_t<n>::min); }
  constexpr static uintn_t<n> max() { return uintn_t<n>(uintn_t<n>::max); }
};

} // namespace std
```

Reply:
Added the comment marking the end of the namespace. We often do not indent for namespace in C++ I believe.